File: gunhoflash/merge-intervals.ts
const MAX_VALUE = 10000;

function merge(intervals: number[][]): number[][] {
  // init up-down counts
  const upCounts = Array(MAX_VALUE + 1).fill(0);
  const downCounts = Array(MAX_VALUE + 1).fill(0);
  for (const [start, end] of intervals) {
    upCounts[start]++;
    downCounts[end]++;
  }

  const answer: number[][] = [];
  let start = 0, end = 0, upDownCount = 0;
  for (let i = 0; i <= MAX_VALUE; i++) {
    const up = upCounts[i];
    const down = downCounts[i];

    if (up !== 0 && upDownCount === 0) {
      start = i;
    }

    upDownCount += up - down;

    if (down !== 0 && upDownCount === 0) {
      end = i;
      answer.push([start, end]);
    }
  }

  return answer;
}

File: gunhoflash/merge-intervals.py
from typing import List

MAX_VALUE = 10000

class Solution:
  def merge(self, intervals: List[List[int]]) -> List[List[int]]:
    # init up-down counts
    upCounts = [0] * (MAX_VALUE + 1)
    downCounts = [0] * (MAX_VALUE + 1)
    for [start, end] in intervals:
      upCounts[start] += 1
      downCounts[end] += 1

    answer = []
    start = end = upDownCount = 0
    for i in range(0, 10001):
      up = upCounts[i]
      down = downCounts[i]

      if up != 0 and upDownCount == 0:
        start = i

      upDownCount += up - down

      if down != 0 and upDownCount == 0:
        end = i
        answer.append([start, end])

    return answer
